### PR TITLE
option nogil=True not required for ufuncs and gufuncs

### LINKED ISSUE #1317: Support nogil=True for ufuncs and gufuncs
```
In [3]: @numba.vectorize(nogil=True)
   ...: def f(a, b):
   ...:     return a + b
   ...:

In [4]: f(1, 2)
KeyError: "Does not support option: 'nogil'"
```

This would be extremely useful for integrating numba ufuncs with dask.array.

### PR DESCRIPTION
**Pull request by pradkrish**
Closes #1317  
nogil=True option is not required for ufuncs and gufuncs. Print a UserWarning if it is provided.

### CODE REVIEW

**Review by gmarkall — CHANGES_REQUESTED** [review]
CUDA ufuncs don't release the GIL, for various reasons (some Python code needed around to setup arguments, kernel launch asynchrony, etc.). When trying to use `nogil=True` an error is produced. For example:

```python
@vectorize(nogil=True, target='cuda')
def f(x, y):
    return x + y
```

Gives:

```
KeyError: "cuda vectorize target does not support option: 'nogil'"
```

This is all fine with the functionality in this PR: However, I think the message in the warning is a little misleading - perhaps "ufuncs and gufuncs always release the GIL with the CPU target" would be more accurate?

**Comment by pradkrish** [comment]
> This is all fine with the functionality in this PR: However, I think the message in the warning is a little misleading - perhaps "ufuncs and gufuncs always release the GIL with the CPU target" would be more accurate?

Thanks for the review, I have modified the warning.

File: numba/np/ufunc/ufuncbuilder.py [hunk 1]
```diff
@@ -82,6 +83,10 @@ def compile(self, sig, locals={}, **targetoptions):
         topt.update(targetoptions)
 
         flags = compiler.Flags()
+        if 'nogil' in topt:
+            warnings.warn("nogil option is not required, ufuncs and gufuncs "
+                          "always releases GIL for CPU target")
```
Comment (stuartarchibald) [inline numba/np/ufunc/ufuncbuilder.py:88]: Suggest using `numba.errors.NumbaWarning` as the warning category such that it's easier for users to filter out things from Numba.

Comment (stuartarchibald) [inline numba/np/ufunc/ufuncbuilder.py:88]:
```suggestion
            warnings.warn("The 'nogil' option is not required, ufuncs and gufuncs "
                          "always release the GIL on the CPU target")
```

File: numba/tests/test_warnings.py [hunk 2]
```diff
@@ -204,6 +205,23 @@ def test_disable_performance_warnings(self):
         out, err = popen.communicate()
         self.assertEqual(popen.returncode, not_found_ret_code)
 
+    def test_warning_vectorize_with_nogil(self):
+        @vectorize(nogil=True)
```
Comment (stuartarchibald) [inline numba/tests/test_warnings.py:209]: Would be good to check `guvectorize` as well given the error message mentions it?

**Review by stuartarchibald — COMMENTED** [review]
Thanks for the PR, just a couple of minor things to resolve else good to go! Thanks again!

**Comment by esc** [comment]
@pradkrish just enquiring about the state of this PR, where is it at?

**Pull request closed without merging**